### PR TITLE
fix: Move all parameters from query to JSON body

### PR DESCRIPTION
**Pull request by provokateurin**
Closes https://github.com/nextcloud/openapi-extractor/issues/88

This fixes a lot of cases where the query parameters was just typed as string because it was not possible to correctly set the type.

### CODE REVIEW

**Review by come-nc — APPROVED** [review]
Tested on security_guard and looks promising.

File: generate-spec
Comment (nickvergessen) [inline generate-spec:46]: When would someone do this? Who would need to do this?

Since it's in `generate-spec` I guess the maintainer of the API providing code, but it seems it's all-or-nothing for the whole app? Should I enable this for Talk?

**Review by nickvergessen — CHANGES_REQUESTED** [review]
Needs at least more documentation

File: generate-spec
Comment (provokateurin) [reply generate-spec:46]: Good question, we should probably just switch all parameters into the body all the time so that the experience is always the same and not a mess of different locations for different parameters when calling the endpoint.

Comment (vitormattos) [reply generate-spec:46]: I made a local test and worked fine to me.

Comment (provokateurin) [reply generate-spec:46]: > OCI endpoint

You mean OCS?

> is necessary to have all parameters specified at requestBody because if not, the documentation will be wrong and will consider that all parameters is send by query string

No that is wrong, parameters can be either in the body or the query parameters. Even putting values in both for a single request works afaik.

> Only make sense to use this when is not a GET because by definition, GET haven't body.

No, GET requests can also have bodies, just some implementations can cause troubles which is not the case for Nextcloud/PHP (see https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET). For the sake of simplicity and consistency all parameters will just go into the body regardless of the HTTP method.

> I made a local test and worked fine to me when I used --parameters-to-body

What are you talking about? I know this code works (and the previous code also works just fine)

Comment (vitormattos) [reply generate-spec:46]: > You mean OCS?

Sorry, is this, OCS. Was a typo.

> For the sake of simplicity and consistency all parameters will just go into the body regardless of the HTTP method.

 Ok

> What are you talking about? I know this code works (and the previous code also works just fine)

Sorry again by my comment at this PR. I'm only helping the code review making local tests.

Comment (provokateurin) [reply generate-spec]: The flag is gone now and everything is moved to the body by default.

**Comment by provokateurin** [comment]
I will still have to do some testing to ensure the updated specs are correct.

**Comment by provokateurin** [comment]
My API unit test suite from https://github.com/nextcloud/neon/tree/main/packages/nextcloud/test confirms that these specs work exactly as intended. From my side this is ready to be merged.

**Review by come-nc — APPROVED** [review]
Does this work fine with `ocs_api_viewer` app?

**Comment by provokateurin** [comment]
It should, but let me give it a try

**Comment by provokateurin** [comment]
Oof, it seems like ocs_api_viewer is broken in that regard because it doesn't send the body at all. This has to be a bug in the library as the body is also not added to the code examples, but I couldn't find an issue about it :thinking:

**Comment by provokateurin** [comment]
After another quick check it makes somewhat sense: GET requests are "not allowed" to have request bodies, so this is likely why it is omitted. This is not really true though, as the spec does not forbid and it is just considered undefined behavior while most web servers just handle it right (See the note at the top of https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET). All other endpoints are fine, I was just unlucky to select a GET endpoint when I quickly tested it (or rather lucky because this way I found the problem).
So we could change this to only move the parameters in case it is not a GET or DELETE operation, but that won't solve the serialization problems that are the driving force behind this PR. I would vote for ignoring this as the changes are required to send valid requests in quite a few cases and the "forbidden" part is not our problem but rather a different interpretation of the spec which is not correct IMO. I would also go and change this in the stoplight/elements library we use in ocs_api_viewer to make this work for us.

@come-nc @nickvergessen what do you think?

**Comment by come-nc** [comment]
> So we could change this to only move the parameters in case it is not a GET or DELETE operation, but that won't solve the serialization problems that are the driving force behind this PR. I would vote for ignoring this as the changes are required to send valid requests in quite a few cases and the "forbidden" part is not our problem but rather a different interpretation of the spec which is not correct IMO. I would also go and change this in the stoplight/elements library we use in ocs_api_viewer to make this work for us.

I’m fine with both solutions, either putting parameters into body only for non-GET requests, or for all of them.
It’s more common to have complex parameters for POST/PUT I suppose, but a GET on a search endpoint might have complex objects for the filter for instance.

**Comment by provokateurin** [comment]
@nickvergessen do you want to do a final review or can I go ahead and merge?

**Comment by nickvergessen** [comment]
Go ahead, still having more than 700 unread emails and this is not a prio for me atm.

**Pull request merged**
